Prototype for opening a fileReader directly to a string

jeremiah-corrado opened the pull request:
This PR adds an `openStringReader` factory function to the IO module which opens a `fileReader` directly to a string.

Under the hood, it creates a memory file, populating it with the string's contents, and then returns a fileReader to the memory file.

- [ ] paratest

twesterhout commented:
@jeremiah-corrado, I tried running my tests on this PR, and they segfault. I think, this is totally unrelated to your PR, but if you could rebase it on current main, I can rerun them again.

twesterhout commented:
@jeremiah-corrado I take it back -- it might be related to your PR after all :smile: I've tried compiling with and without your latest commit, and my code segfaults with it. I have code blocks like this one:
```chapel
    proc from_json(s : c_ptrConst(c_char)) throws {
      const c_str = ls_hs_operator_from_json(s);
      defer ls_hs_destroy_string(c_str);

      var file = openMemFile();
      var writer = file.writer();
      writer.write(string.createBorrowingBuffer(c_str));
      writer.close();
      var r = file.reader(deserializer=new jsonDeserializer());
      var p = r.read(HsResult(uint(64))).value;
      return p:c_ptr(void):c_ptr(ls_hs_operator);
    }
```
Could the new `_file_handle` member be causing a double close or something?

jeremiah-corrado commented:
Okay, thanks for the feedback! I think a double-free of some kind would definitely make sense (although I'm not sure why that didn't come up with the test I added in this PR). I'll dig into this and see if I can fix the segfault. 

Also, not sure what happened with that rebase 😬, I'll see if I can fix that as well.

vasslitvinov commented on modules/standard/IO.chpl:
```diff
@@ -5428,6 +5428,72 @@ private proc openReaderHelper(path:string,
+  var f = openMemFile(),
+      w = f.writer();
+  w.write(slocal);
+  w.close();
```
@jeremiah-corrado I would expect `w.write(s)` to localize automatically as needed. For example, this works, or is it just me?
```chpl
use IO;
var f = open("vass", ioMode.cw);
var w = f.writer();
on Locales[2] {
  var s = "my string\n";
  on Locales[0] {
    w.write(s);
  }
}
```

Also I expect that a `const` intent is more appropriate here, whether you localize explicitly or not. This function merely reads the string, so there is no need to create a copy of the actual argument [when the actual is not dead].

jeremiah-corrado commented:
Thanks for the feedback @vasslitvinov! I agree with what you said and added tests to show that the multi-locale case is working for `openStringReader` and `openBytesReader`. I also changed the formal intent to `const`.

(This involved fixing a localization bug in `readLine`. There are a few other such bugs I noticed in the IO module — I intend to fix those in a separate PR).

vasslitvinov commented on test/library/standard/IO/stringBytesDirectReader/COMPOPTS:
```diff
@@ -0,0 +1 @@
+--warn-unstable
```
Nit: if possible, add a newline here.

bradcray commented:
Nice to see this go in!
